iotilecore/iotile/core/hw/debug/firmware_image_analyzer.py

```diff
@@ -0,0 +1,150 @@
+from iotile.core.utilities.intelhex import IntelHex
+from iotile.core.exceptions import *
+from tempfile import NamedTemporaryFile
+import os
+import subprocess
+import struct
+
+class FirmwareImageAnalyzer:
+    """A firmware image analyzer using IntelHex"""
+
+    KNOWN_FIRMWARE_NAMES = ['boot52', 'NRF52 ', 'firmPP', 'UART  ', 'firmFF']
+    SIZE_OF_CDBAPPINFO = 32
+    MAGIC_NUMBER_BYTE_OFFSET = 24
+
+    def __init__(self, firmware, magic_number=0xBAADDAAD):
```

Comment:
I would recommend making this class a bit more immutable so that it's easier to work with:

- You already have a `magic_number` option in `__init__`, so there's not a need to allow the magic number to be set later.   The user can just create the class when once they know the magic number they want to use.
- This means all of the analysis can be done immediately in `__init__` since the possibility of the magic number changing was the only thing I can see that lead to deferring the analysis until later.
- This means that the class can just have a declarative `app_info` property that contains the results.  If analysis wasn't possible, it can raise a `DataError` from the `__init__` method.

This should simplify the class and remove some edge cases that currently are broken.  For example, you are allowed currently to call `set_firmware_magic_number` after initial construction but if you call it after the first time you have called any accessor function, it will be silently ignored because the results will have already been cached.

Moving everything to happen during `__init__` makes it more clear.  You create the class to analyze a firmware and then you can access the results.  If you want to analyze the firmware differently, create another instance of the class with different arguments.

Reply:
Fixed in next push.